### PR TITLE
Begin adding WASI support

### PR DESCRIPTION
**Pull request by MarkMcCaskey**
Very much a work in progress.  Probably will require updating more of the existing code toward `ImportObject` and away from existing imports, though I'm not sure.

This PR is not in a state where it can be merged and won't be until we ship WASI support in the C API.  Before doing so we want to verify, by implementing it in a language extension, that the updates to the C API are sufficient.

### CODE REVIEW

**Comment by Hywan** [comment]
Thanks!

Can you rebase on top of `master` please? Since #73, Wasmer 0.8.0 is used, which will be easier for everyone.

File: wasmer/import.go [hunk 1]
```diff
@@ -14,6 +14,139 @@ type ImportedFunctionError struct {
...
+	return &ImportObject{inner}
+}
+
+
+// Creates an `ImportObject` with the default WASI imports.
+// Specify arguments (the first is the program name),
+// environment variables ("envvar=value"), preopened directories
+// (host file paths), and mapped directories (host file paths with an
+// alias, see `MapDirEntry`)
+func NewWasiImportObject(args []string, envs []string,
```
Comment (syrusakbary) [inline wasmer/import.go]: I think might be a good idea to move the WASI specific stuff to a different file: `wasi.go`

File: wasmer/instance.go [hunk 2]
```diff
@@ -167,11 +167,23 @@ func newInstanceWithImports(
...
 
-	var wasmExports *cWasmerExportsT
-	var hasMemory = false
+	return Instance{instance: instance, imports: imports, Exports: exports, Memory: &memory}, nil
+}
 
+// Returns the exports, whether it has memory or an error
+func getExportsFromInstance(instance *cWasmerInstanceT, memory *Memory) (map[string]func(...interface{}) (Value, error), bool, error) {
+	var exports = make(map[string]func(...interface{}) (Value, error))
+	var hasMemory = false
```
Comment (syrusakbary) [inline wasmer/instance.go]: I think `hasMemory` might be better calculated in a function apart?

Comment (syrusakbary) [reply wasmer/instance.go]: Or rather than in a function apart, just checking if in the returned `Instance` , the memory is null?

Comment (MarkMcCaskey) [reply wasmer/instance.go]: Yeah, that's reasonable -- I just kept the existing structure when I copied and pasted that inner part out; checking for null should work though!

**Comment by robinvanemden** [comment]
How far along is this branch? We currently have completed a machine learning C++ application that compiles nicely using both clang and emscripten standalone, and now runs on both Wasmer and WAVM. But it fails to instantiate the wasm using go-ext-wasm, dueto some missing WASI imports. Would be super if we can make this work, as it would be the final step towards completing our framework.

**Comment by Hywan** [comment]
I'm literally working on this right now. Expect to be merged in a couple of hours if everything goes smoothly.

**Comment by robinvanemden** [comment]
Incredible! The timing couldn't be any better. Thanks!

File: wasmer/test/testdata/wasi_hello_world.wat [hunk 3]
```diff
@@ -0,0 +1,33 @@
+;; Copied from https://github.com/CraneStation/wasmtime/blob/master/docs/WASI-tutorial.md and then modified
+;; Licensed under Apache 2.0; see https://github.com/CraneStation/wasmtime/blob/master/LICENSE
+(module
+    (import "wasi_unstable" "fd_write" (func $fd_write (param i32 i32 i32 i32) (result i32)))
```
Comment (Hywan) [inline wasmer/test/testdata/wasi_hello_world.wat]: It must be `wasi_snapshot_prevew1` now.

**Comment by Hywan** [comment]
Review feedbacks fixed by https://github.com/wasmerio/go-ext-wasm/commit/c07c1017f9ebede0af5241622116be8367bbc1f6, https://github.com/wasmerio/go-ext-wasm/commit/b91099d08b5a069fe1d13edf3d8474f01778859b and https://github.com/wasmerio/go-ext-wasm/commit/bf837c2dd588e346ecafe9a6c1cc956801d04626 (the last one fixes WASI itself, since the namespace has changed from `wasi_unstable` to `wasi_snapshot_preview1`.

Thanks!

**Comment by Hywan** [comment]
The test prints `hello world` as expected, but I wish we can find a solution to test the output to avoid regression in the future.

**Comment by robinvanemden** [comment]
Hi Hywan! Thanks for all of the work you have been putting in. I have been testing the latest commit, and still get the following:

`7 link errors: (1 of 7) Import not found, namespace: wasi_unstable, name: proc_exit (2 of 7) Import not found, namespace: wasi_unstable, name: environ_get (3 of 7) Import not found, namespace: wasi_unstable, name: environ_sizes_get (4 of 7) Import not found, namespace: wasi_unstable, name: fd_write (5 of 7) Import not found, namespace: wasi_unstable, name: fd_seek (6 of 7) Import not found, namespace: wasi_unstable, name: fd_read (7 of 7) Import not found, names pace: wasi_unstable, name: fd_close`

Does that make sense? Does it have anything to do with the wasi_snapshot_preview1 style import?

**Comment by robinvanemden** [comment]
Wait, I probably need to use wasm.NewDefaultWasiImportObject()! I will test this later. Thanks again for all of your work on this!

**Comment by robinvanemden** [comment]
Super! Using `NewDefaultWasiImportObject` did the trick: I can now use the WASI ABI in go-ext-wasm. There is one catch though. To make an existing WASI run (compiled with either the latest clang, WasiENV, or Emscripten) I now need to find&replace `wasi_unstable` with `wasi_snapshot_preview1`. This conversion is not necessary when I run the same WebAssembly by calling Wasmer directly. Apparently, Wasmer supports both import namespaces that way. Is there some elegant way to do the same in go-ext-wasm?

**Pull request merged**